### DISCUSSION TITLE
What is the approach for non-POCOs in 4.X

### DISCUSSION THREAD

**Discussion by SebastianStehle** [dotnet/orleans#7569] [Q&A]
Hi,

The Release describes what to do with POCOs. But what is the approach for Value types and collections and everything else that is not a POCO?

I understand how POCOs get serialized. You loop over all the properties and then you have default serializers for the most common value types (DateTime, Date, Guid and so on) and collections (List, Dictionary, HashSet ...).

But what is, when I have types, that I cannot annotate, for example a custom value type for a Guid, or NodaTime. In 3.X it works, because the underlaying fields are often serialized (in my understanding). But what do I have to do now? Is there a new interface for a custom converter / serializer?

**Comment by ReubenBond** [comment]
Hi Sebastian, could you elaborate?

**Reply by SebastianStehle** [reply]
I updated my question

**Comment by ReubenBond** [comment]
A surrogate can be used in place of the original type. Below is an example of creating a serializer for a type which cannot be annotated because it comes from an assembly which you don't own.

The general pattern is to create a surrogate type, have Orleans generate a serializer for that, and then create a converter. In the example below, I've also defined a copier. Often times, the type is immutable, so the copier can simply return the input.

``` C#
[RegisterSerializer]
public sealed class UriCodec : GeneralizedReferenceTypeSurrogateCodec<Uri, UriSurrogate>
{
    private readonly TypeConverter _uriConverter = TypeDescriptor.GetConverter(typeof(Uri));

    public UriCodec(IValueSerializer<UriSurrogate> surrogateSerializer) : base(surrogateSerializer)
    {
    }

    public override Uri ConvertFromSurrogate(ref UriSurrogate surrogate) => surrogate.Value switch
    {
        null => null,
        _ => (Uri)_uriConverter.ConvertFromInvariantString(surrogate.Value)
    };

    public override void ConvertToSurrogate(Uri value, ref UriSurrogate surrogate) => surrogate = value switch
    {
        null => default,
        _ => new UriSurrogate
        {
            Value = _uriConverter.ConvertToInvariantString(value)
        },
    };
}

[GenerateSerializer]
public struct UriSurrogate
{
    [Id(1)]
    public string Value { get; set; }
}

[RegisterCopier]
public sealed class UriCopier : IDeepCopier<Uri>, IGeneralizedCopier
{
    public Uri DeepCopy(Uri input, CopyContext context) => input;

    // Uri has derived types and these are handled by declaring this as a "generalized copier" which
    // knows how to copy them. Usually, you don't need to do this. The "System.Type" type also needs something similar
    public object DeepCopy(object input, CopyContext context) => input;
    public bool IsSupportedType(Type type) => typeof(Uri).IsAssignableFrom(type);
}
```

Note that if the type lives in an assembly which you do own, then you have the option of annotating the type using the correct attributes but generating code in a separate assembly. In order to accomplish that, the assembly with the types should *not* have the code generator installed (only Microsoft.Orleans.Serialization.Abstractions), and another assembly should have both the code generator installed as well as an attribute in the form of `[assembly: GenerateCodeForDeclaringAssembly(typeof(SomeTypeInTheOtherAssembly))]`.

**Reply by SebastianStehle** [reply]
it looks like a lot of code. Is there no shortcut for it? e.g. a lot of value types support type conversion. A serializer that supports that could work in a lot of cases.

**Reply by ReubenBond** [reply]
Could you expand on what you mean by type conversion here, how would it look in code? Are you suggesting that the process explained above could be more automated by having just a surrogate type (with no hand-coded conversion serializer) which has conversion operators to/from the target type?

There's also the option of plugging another serializer in, eg there is a support package for Newtonsoft.Json which accepts a `Func<Type, bool>` to indicate which types should be serialized using that. A similar technique can be used for other serializers.

**Reply by SebastianStehle** [reply]
I like value objects and I think I should use them more often. But one annoying part is that you have to write so many serializers. e.g. if you write a Web API with Orleans and MongoDB you need 3 serializers per type.

This could be solved with code generation I guess but because most value types have a primary representation I use the following approach:

```
[TypeConverter(typeof(DomainIdConverter)]
public record struct DomainId {
   private readonly string value;
}

// e.g. JSON
public sealed class TypeConverterJsonConverter<T> : JsonConverter, ISupportedTypes
{
	private readonly TypeConverter typeConverter = TypeDescriptor.GetConverter(typeof(T));

	public override object? ReadJson(JsonReader reader, Type objectType, object? existingValue, JsonSerializer serializer)
	{
		return typeConverter.ConvertFromInvariantString(reader.Value?.ToString()!);
	}

	public override void WriteJson(JsonWriter writer, object? value, JsonSerializer serializer)
	{
		writer.WriteValue(typeConverter.ConvertToInvariantString(value));
	}
}
```

Of course this example is simplified. But if the serializer would support a concept of serializer factories or conventions we could easily write serializers for many systems.

For example I also use surrogates, so I have a simple interface for that:

```
public interface ISurrogate<T>
    {
        void FromSource(T source);

        T ToSource();
    }
```

With such a serializer factory you can just scan for types and register them automatically.

**Reply by ReubenBond** [reply]
Sounds very reasonable to me. So if we had something like this, would that suffice?
``` C#
public interface ISurrogate<T> // could optionally have a TSurrogate "self-type" parameter and a static FromSource method
{
    void FromSource(T source);
    T ToSource();
}

[RegisterSurrogate]
public class MySurrogate : ISurrogate<MyPoco>
{
        void FromSource(MyPoco source) => ...
        MyPoco ToSource() => ...
}
```

**Reply by SebastianStehle** [reply]
These are just two examples. I would prefer to have factories as in System.Text.Json: https://docs.microsoft.com/en-us/dotnet/standard/serialization/system-text-json-converters-how-to?pivots=dotnet-6-0#steps-to-follow-the-factory-pattern

Then I can implement my own code / converter / serializer.

**Reply by SebastianStehle** [reply]
Any updates?

**Reply by ReubenBond** [reply]
I hope to get to this soon, but I don't have any updates yet

**Reply by ReubenBond** [reply]
What would you like your custom serializer to look like? You can of course already implement your own serializer today, but it's complicated (it's intended for generated code). We could possibly implement a lower-performance scheme using named properties, similar to JSON... or you could use System.Text.Json if you want. Would you like that?

**Reply by SebastianStehle** [reply]
I would.prefer the convention or factory based approach. The serializer itself is okay, I just do not want annotations everywhere.

**Reply by ReubenBond** [reply]
I don't quite understand: how would it look? Do you mind offering a proposal?

**Reply by ReubenBond** [reply]
Would a serializable `ISurrogate<T>` interface like the one you described above be sufficient? Otherwise, could you please elaborate? I'd like us to address this.

**Reply by SebastianStehle** [reply]
I can describe it tomorrow.

**Reply by VincentH-Net** [reply]
@ReubenBond  A significant % of Orleans 4 devs are likely to hit this limitation to POCO and owned types, early on.

I encountered this on the very first best practice pattern I applied to my very first Orleans 4 grain:
- Best practice (performance rule): return error codes from grains for expected errors (e.g. validation) instead of raising exceptions
- Use the most well known lightweight NuGet for that pattern [FluentResults](https://www.nuget.org/packages/FluentResults/), > 1 million downloads
- Grain methods returning a `FluentResults.Result` e.g. `Task<FluentResults.Result<UserInfo>> GetUser(Guid id);` where `UserInfo` is an owned `record` type with `[GenerateSerializer]` on it, fail with exception `Orleans.Serialization.CodecNotFoundException: Could not find a copier for type FluentResults.Result`



It may not be many types per application, but if will be many devs.
The `ISurrogate<T>` approach suggested above still is a lot of work to code manually - replicate the entire data structure and copy that, and maintain on update of external type.
 
Would it be possible to let the attribute accept a type parameter for types defined in 3rd party assemblies, and then all the glue code gets generated? So an assembly level attribute:
```csharp
[GenerateSerializer(typeof(FluentResults.Result)]
```

But even better would be to just do all this automatically; the fact that a type needs serialization somewhere could triggers the serializer codegen, regardless of whether it is POCO, owned, or not owned. 
Then there is no limitation to serialization. which seems odd anyway for a mature product that aims to make life easier for C# devs (ofc I say that with the benefit of blissful ignorance about how hard this may be to solve)

**Reply by ReubenBond** [reply]
The main problem is version tolerance. A declaration like `[GenerateSerializer(typeof(FluentResults.Result)]` doesn't tell us anything about how the type should be serialized (what properties/fields it contains). If the type changes, the auto-generated property identifiers might not match (meaning it would fail in difficult to predict ways).

If you were using JSON over HTTP ("REST") or gRPC, you have the same problem. We could declare that *System.Text.Json* is the default method for serialization for when a developer hasn't specified another approach. I'm ok with that, if that's what everyone wants. There's a PR adding support for System.Text.Json here: https://github.com/dotnet/orleans/pull/7756. The biggest issue I see with that is that it won't solve all problems (eg, exceptions) and STJ doesn't support polymorphism well (eg, `FluentResults.Result` has `IResult<T>` and `ResultBase` - supporting those interface/abstract base types requires supporting polymorphism).

Another alternative, if people want unreliable magic, is to implement something like the `ILBasedSerializer` which was present in 3.x: it would use reflection + IL to create a serializer automatically. It wouldn't be version tolerant, so you may see failures when migrating from one library/framework version to another (we see this from time to time)

The magic serialization approach is nice for getting started, but it's also a trap.

**Reply by VincentH-Net** [reply]
Thanks for the clear explanation @ReubenBond ! I understand that this is a choice between ease of use and safety.

I can think of a few additional considerations:
- Is unversioned serialization of grain method parameters/return values less unsafe than in other places e.g grain state or streams? It that is so, maybe we can relax the 'forced safety' approach only for grain calls?
- Developer satisfaction can perhaps be improved with a prominent explanation of the why in the error message (link). Your answer above had that effect on me at least 🙂
- Safe by default is good, but having no choice to avoid extra coding can be frustrating in cases where the dev can make other choices to avoid the risk (e.g. ensuring no (breaking) updates occur). Perhaps an explicitly 'unsafe/unversioned' named serialization attribute? Or add an (required?) assembly version parameter?

**Reply by SebastianStehle** [reply]
I was thinking about it again and I think the ISurrogate<T> interface is good enough. This means you can solve all serialization issues outside of your business layer.

Custom serializer still exist, right? Because I use JSON for Orleans 3.

**Reply by VincentH-Net** [reply]
@ReubenBond wrote:
> Sounds very reasonable to me. So if we had something like this, would that suffice?
> 
> ```cs
> public interface ISurrogate<T> // could optionally have a TSurrogate "self-type" parameter and a static FromSource method
> {
>     void FromSource(T source);
>     T ToSource();
> }
> 
> [RegisterSurrogate]
> public class MySurrogate : ISurrogate<MyPoco>
> {
>         void FromSource(MyPoco source) => ...
>         MyPoco ToSource() => ...
> }
> ```

@SebastianStehle wrote:
> I was thinking about it again and I think the ISurrogate interface is good enough. This means you can solve all serialization issues outside of your business layer.

I second that, if above code would enable `T` serialization transparently and it would support any type.
E.g. you can just do:
```csharp
[RegisterSurrogate]
public class MySurrogate : ISurrogate<MyAnyType> { ... }

public interface IMyGrain
{
    Task<MyAnyType> Get();
}
```

**Reply by VincentH-Net** [reply]
Another thought @ReubenBond:

It seems like there are several type categories in how Orleans 4 handles serializing:

1) User owned types.
  These types are by far the most likely to change over the lifetime of an application and this is where versioning is most valuable, and they require annotation and controlling the versioning with id numbers

2) Foundational types.
  E.g. types of the .NET framework; it seems that these types are assumed not to change wrt serialization, so they do not need explicit Orleans versioning with id numbers

Most cases for this issue relate to non-owned, non-foundational types so their handling is unsupported.

**What if** we let the developer explicitly indicate which types are considered foundational/unchanging wrp serialization _in the context of their application_, so no versioning will be used for serializing them? E.g.
```csharp
[GenerateUnversionedSerializer(typeof(AnyType))
```

You could include a (mandatory?) 'string assemblyVersion` parameter, validated against [Version](https://docs.microsoft.com/en-us/dotnet/api/system.version), that can be included in the serialized instance and enforced at codegen time (so it won't compile if the version of the type's containing assembly does not match - the compile error can again contain / link to a clear description of the risk involved in updating). This will prevent unwanted updates at compile time and give a clear exception at runtime (indicating the type name, payload assembly version and serializer assembly version) if a dev makes a wrong choice. E.g.
```csharp
[GenerateAssemblyVersionedSerializer(typeof(AnyType), "1.1.0")
```
 
It could be implemented with  the approach you described above:
> Another alternative, if people want unreliable magic, is to implement something like the `ILBasedSerializer` which was present in 3.x: it would use reflection + IL to create a serializer automatically. It wouldn't be version tolerant

This way, there is safety out of the box but the developer also has an explicit choice in the context of their application - not **everything** needs to be serialized with versioning at all costs in every application.

I would prefer this over the `ISurrogate<T>` approach.

Best of both worlds?

**Reply by SebastianStehle** [reply]
I know now what I was missing with the ISurrogate approach and serialization in general: The ability to write generic serializers.

In my application I have a lot of types that are just string representations:

* DomainId (something like Guid)
* Instant
* UserId
* ...

Of course I can write serializers for JSON, database and Orleans, but it is annoying. What I usually do, is to write a type converter for them and then use this approach in all serializers. E.g. lets consider I have a type converter, I could write my serializer or surrogate like this:

```
public sealed class StringSurrogate<T> : ISurrogate<T>
{
        private static readonly TypeConverter typeConverter = TypeDescriptor.GetConverter(typeof(T));
        
        public string T Value { get; set; }
        
        void FromSource(T source)  {
           Value = source.ToString();
        }

        T ToSource() => ...
                return (T)typeConverter.ConvertFromInvariantString)!;
}
```

But then we need to register the surrogates somehow. One way would be just a method somewhere to register surrogates or a converter factory.

I assume with your serializer you can write custom converters if you really need the performance (e.g. for custom collections).

Then you could introduce an interface like this

```
interface IConverterFactory
{
   Converter? Create(Type targetType);
}
```

You can then either create a converter manually or you return a surrogate converter.

```
class StringConverters : IConverterFactory
{
   Converter? Create(Type targetType)
   {
      if (targetType.GetAttributes<ConverterAttribute>() != null)
      {
         var surrogateType = typeof(Surrogate<>).MakeGenericType(targetType);
         
         return Converter.WithSurrogate(targetType, surrogateType);
      }
      
      return null;
   }
}
```

**Reply by ReubenBond** [reply]
Would JSON (eg, System.Text.Json) suffice for all of these cases?
If so, then we can make the System.Text.Json support in Orleans as efficient as possible, removing any additional allocs, etc - currently the implementation is fairly efficient (using pooled buffers).

**Reply by SebastianStehle** [reply]
I am not sure I get that. My understanding was that it works like this:

1. Each parameter or return type is serialized independently using a serializer (or codec in 4.0).
2. The default codec is the new version tolerant serialization format.

So the question is: How much flexibility do you want in the new format? If it is faster then it could be valuable to provide the opportunity to write custom converters, either with manual converters or surrogates. If you do not want this flexibility I can live with json or any other custom codec.

**Reply by ReubenBond** [reply]
Your understanding is correct. The new format supports embedding other codecs (any general-purpose codec, such as JSON, etc).
The native format is faster than System.Text.Json, but probably not by that much: the main goal of it is version tolerance and high-fidelity support for .NET types. It has some support for surrogates in some form (implementing `ReferenceTypeSurrogateCodec`/`GeneralizedValueTypeSurrogateCodec`). Eg: https://github.com/dotnet/orleans/blob/506fb2f5a5474f6a3b05304aec77d1b206b89ede/src/Orleans.Serialization/Codecs/ImmutableArrayCodec.cs#L13.

It also has some support for something like the `IConverterFactory` that you mention, via `ISpecializableCodec`: https://github.com/dotnet/orleans/blob/506fb2f5a5474f6a3b05304aec77d1b206b89ede/src/Orleans.Serialization/Serializers/IGeneralizedCodec.cs#L22

I'm not against making usability improvements (eg a layer on top of what's already there), but I wonder if "just use JSON/etc" is the right approach for most of these cases. We could use an interface to indicate whether or not JSON is supported for a given type, instead of passing a delegate.

**Reply by SebastianStehle** [reply]
Okay, that's cool. Perhaps we should just close the discussion here and open it when 4.0 is released. I have only worked with 4.0 to migrate the dashboard.

**Reply by ReubenBond** [reply]
I'll open a PR to add `IConverter<TValue, TSurrogate>` and `[RegisterConverter]` soon, and follow it with a better solution for allowing types. I think that will cover most cases.

Example of `IConverter<TValue, TSurrogate>`:
``` C#
[GenerateSerializer]
public struct MyForeignLibraryTypeSurrogate
{
    [Id(0)]
    public int Num { get; set; }

    [Id(1)]
    public string String { get; set; }

    [Id(2)]
    public DateTimeOffset DateTimeOffset { get; set; }
}

[RegisterConverter]
public sealed class MyForeignLibraryTypeSurrogateConverter : IConverter<MyForeignLibraryType, MyForeignLibraryTypeSurrogate>
{
    public MyForeignLibraryType ConvertFromSurrogate(in MyForeignLibraryTypeSurrogate surrogate)
        => new(surrogate.Num, surrogate.String, surrogate.DateTimeOffset);

    public MyForeignLibraryTypeSurrogate ConvertToSurrogate(in MyForeignLibraryType value)
        => new() { Num = value.Num, String = value.String, DateTimeOffset = value.DateTimeOffset };
}
```

**Reply by VincentH-Net** [reply]
@ReubenBond I like the above, thx! Wrt:
> and follow it with a better solution for allowing types

perhaps something like what I described [above](#discussioncomment-2876684) could help?
```csharp
[GenerateAssemblyVersionedSerializer(typeof(AnyType), "1.1.0")
```
(repeated in case it got lost in the threads)

**Reply by ReubenBond** [reply]
> perhaps something like what I described [above](https://github.com/dotnet/orleans/discussions/7569#discussioncomment-2876684) could help?

``` C#
[GenerateAssemblyVersionedSerializer(typeof(AnyType), "1.1.0")
```

That would be a whole new mechanism and I'm not sure yet how it would work: would it break compilation when the target assembly's version changes? If it does change and the developer bumps the version in the attribute, what would the behavior be? Assembly versions are generally unreliable.

We could have something like `[assembly: AllowType(typeof(X))]` and `[assembly: AllowTypesFromAssembly(typeof(X))]`. I was thinking of the lower-level mechanism: adding a more user-friendly interface for checking individual types. I'm considering whether or not we should allow all types by default: the new serializer is simply unable to serialize/deserialize/copy arbitrary/unknown types, and nested codecs must be responsible for their own type loading logic.

**Reply by VincentH-Net** [reply]
> how it would work: would it break compilation when the target assembly's version changes?

Yes; I outlined how it could work [above](https://github.com/dotnet/orleans/discussions/7569#discussioncomment-2876684):
> include a (mandatory?) 'string assemblyVersion` parameter, validated against [Version](https://docs.microsoft.com/en-us/dotnet/api/system.version), that can be **included in the serialized instance** and **enforced at codegen time** (so it won't compile if the version of the type's containing assembly does not match - the compile error can again contain / link to a clear description of the risk involved in updating). This will prevent unwanted updates at compile time and give a clear exception at runtime (indicating the type name, payload assembly version and serializer assembly version) **if a dev makes a wrong choice**

> If it does change and the developer bumps the version in the attribute, what would the behavior be?

Bumping the version in `[GenerateAssemblyVersionedSerializer]` would indicate a breaking change in the serialization of the type, so the grain interfaces that include this type must be versioned as well, and their versions must be bumped at the same time. The configured `GrainVersioningOptions` and deployment approach will then handle things as usual - **if** the change in serialization is backwards compatible (unless Orleans has a means to handle non-backwards compatible grain versions - is there such a thing as a custom version selector strategy?). 

By adding a `ushort serializationVersion` to `[GenerateAssemblyVersionedSerializer]` it is possible to have the relevant grain interfaces version increase automatically when there is a breaking change in the serialization of an external type:
```csharp
static class ExternalTypeVersions
{
    public const ushort MyExternalType = 1;
}

[assembly: GenerateAssemblyVersionedSerializer(typeof(MyExternalType), "1.1.0", ExternalTypeVersions.MyExternalType)

[Version(1 + ExternalTypeVersions.MyExternalType)]
public interface IMyGrain : IGrainWithStringKey
{
    // Signature(s) that contain MyExternalType
}
```
The developer only needs to increase `ExternalTypeVersions.MyExternalType` when updating the dependency and while doing so concluding that there is in fact a breaking change in the serialization of that type (should be guarded by unit tests). So the `assemblyVersion` parameter for the `GenerateAssemblyVersionedSerializer` attribute can be updated while the `serializationVersion` parameter remains unchanged.

So `[GenerateAssemblyVersionedSerializer)` can add safety both at compile time and runtime. 
My thinking was that it would be more safe than simply allowing an external type without any versioning as `[assembly: AllowType] ` would do. Even though assembly versions are not fully reliable, it will be better than nothing and provide a middle ground.

It would be a new mechanism; perhaps we even want all of the above so we have three safety levels for serialization:
- High: `[GenerateSerializer]` for owned types and surrogates
- Medium: `[GenerateAssemblyVersionedSerializer]`
- None: `[AllowType]`

Btw instead of separate attributes we could us a single one for external types, and also support generating serializers for many types with a single attribute, by adding an additional parameter to match against the type names in the assembly of the specified type:
```csharp
[assembly: GenerateExternalTypeSerializer(typeof(MyExternalType), "1.1.0", 1)]                // Single external type, assembly versioned
[assembly: GenerateExternalTypeSerializer(typeof(MyExternalType), "1.1.0", 1, "Namespace.*")] // External types in namespace within assembly, assembly versioned
[assembly: GenerateExternalTypeSerializer(typeof(MyExternalType))]                            // Single external type, unversioned
[assembly: GenerateExternalTypeSerializer(typeof(MyExternalType), "*")]                       // All external types in assembly, unversioned
```

I would be happy with just supporting unversioned external types; however also supporting assembly versioned external types may be a useful middle ground; it would offer devs a measure of safety while not needing to write a surrogate for every external type they use; especially if that are many types like domain objects created by another team.

**Reply by ReubenBond** [reply]
Interesting idea, @VincentH-Net. It would require the code generator to inspect the foreign assembly and assign ids to the properties/fields in it. I think it's an interesting idea, but I'm concerned that it's too magic and too difficult to get right, since downstream developers don't know what the upstream developers have changed.

Additionally, if you were ever storing any of these types, then the version number wouldn't stop things from blowing up since you would still encounter cases where field orders have changed. Ultimately, the version-number approach could work, but I think it would require a schema repository, eg checked into source control.

I've opened a PR for surrogates here: https://github.com/dotnet/orleans/pull/7776

**Reply by VincentH-Net** [reply]
Valid points, thx!

- I had not thought about the likelihood of those types being stored as well. This makes it likely that issues will not be prevented - so the added security when compared to [AllowType] would be less than I thought.
- Maintaining a schema repo reduces/nullifies the effort saved when compared to creating surrogates

All in all I agree that the idea is not likely to add enough value to justify adding the mechanism (and educating devs about it).

Thx for mentioning me in the surrogate PR; @ReubenBond how could I keep track of below?
> and follow it with a better solution for allowing types

I'd like to follow `[assembly: AllowType(typeof(X))]` and `[assembly: AllowTypesFromAssembly(typeof(X))]` or equivalent

**Reply by ReubenBond** [reply]
Part of it is here: https://github.com/dotnet/orleans/pull/7781
I'm yet to implement those attributes, but I don't think they're needed anymore: any serializer will now have the ability to allow a given type to be serialized by implementing ITypeFilter

**Comment by SebastianStehle** [comment]
Btw: What is the approach for Exceptions? The old ISerializable thing?

**Reply by SebastianStehle** [reply]
Any updates?

**Reply by ReubenBond** [reply]
For exceptions, you will be able to tell Orleans to serialize the otherwise-unserializable exception via `ExceptionSerializationOptions`, either by adding a namespace prefix or by setting a filtering delegate. It will not serialize the properties on the exception in that case, though: only the properties on the base type (stack, message, type, etc)

**Reply by SebastianStehle** [reply]
So you cannot use something like ValidationException anymore?

**Reply by SebastianStehle** [reply]
They are dozens of exceptions where I would be interested in the other properties as well.

**Reply by ReubenBond** [reply]
That's correct: you wouldn't be able to sensibly use `ValidationException` unless you wrote a custom serializer for it. Perhaps we can add certain exception types in-the-box.

The priority is to enable version-tolerance. I'd say we could use `ISerializable` to support it, but [as you can see here](https://github.com/dotnet/runtime/blob/e6563049a2e88d7eb92efa0ba8cfb88c763ea06a/src/libraries/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/ValidationException.cs#L77), ValidationException does not implement it correctly.

I'm open to your suggestions.

**Reply by SebastianStehle** [reply]
Okay the example with ValidationException was stupid, because I have my own ValidationException.

There are a lot of exception types that have custom properties and I would say it should be a best practice. Something like a database error code can be handled with a property easily and it should not be necessary to parse the message to get the result.

I would recommend the following:

1. Support SerializationInfo in the serializer. It is basically just a dictionary, so it should not be too complicated.
2. If an exception supports ISerializable serialize the exception.
3. If the exception cannot be serialized use a wrapper exception and try to obtain as much infos from the original exception as possible. Perhaps with reflection, json serializer or whatever.

For me it happened a few times that I lost the information from an exception because it could not be serialized. I have therefore written a filter for that: https://github.com/Squidex/squidex/blob/master/backend/src/Squidex.Infrastructure/Orleans/ExceptionWrapperFilter.cs.

This is an typical error that often happens in production only, because locally or on the CI machine you will not see it if you do not run a cluster. Therefore we should avoid to loose error details.

I understand and support the idea to move away from the "It works magically until it does not work" approach but exceptions are another story. When you have a look to a method or grain signature you see the input and output. But exceptions do not follow this rule. Some exceptions just flow from an class trough a grain to asp.net core middleware. It is very easy to forget to write a serializer for such an exception.

**Reply by ReubenBond** [reply]
> Support SerializationInfo in the serializer. It is basically just a dictionary, so it should not be too complicated.

We have some support for that already, it's not too difficult. It will work as long as the inner types are serializable, too.

> If an exception supports ISerializable serialize the exception.

Sure, we can do that.

> If the exception cannot be serialized use a wrapper exception and try to obtain as much infos from the original exception as possible. Perhaps with reflection, json serializer or whatever.

We can special case `Exception` to use a fallback if the original type is not serializable, but I don't want to introduce the fragility of reflection or attempt to fall back to JSON serialization when the developer hasn't configured it.

Most exceptions are not designed with remoting in mind, which unfortunately leads to things like an HttpException which carries the whole world with it. Using reflection to try to maintain fidelity for these exceptions often results in additional problems, so I think it's the wrong solution.

**Comment by oising** [comment]
What about using a CustomReflectionContext ( https://docs.microsoft.com/en-us/dotnet/api/system.reflection.context.customreflectioncontext?view=dotnet-plat-ext-6.0 ) to synthesize the [GenerateSerializer] and [Id] attributes needed? This would allow Hagar to remain mostly unmolested -- you'd have to provide an way to provide an optional reflection context -- while allowing a configuration - maybe a manifest config text file - which explains how to annotate external types for which you don't have the source.

**Reply by ReubenBond** [reply]
The code generator inspects the assemblies/code via Roslyn at build time, so `CustomReflectionContext` is not available/applicable. If having external annotations (eg on a dummy mirror type or possibly in some IDL file) is satisfactory, then something like that might be an option.

I'm considering other options, too, like a more user-friendly way to construct serializers programmatically, eg using a builder pattern to map names to fields/properties.

**Comment by koenbeuk** [comment]
Related question, are there plans on properly supporting record types? Currently we'll have to define al Members explicitly and mark them with an Id attribute, e.g.
```csharp
[GenerateSerializer]
record Person(string FirstName, string LastName) {
    [Id(0)]
    public string FirstName { get; init; }

    [Id(1)]
    public string LastName { get; init; }

    [Id(2)]
    public string? MiddleName { get; init; }
}
```

This could hypothetically be simplified to something in the lines of:
```csharp
[GenerateSerializer]
record Person(string FirstName, string LastName) {
    [Id(0)]
    public string? MiddleName { get; init; }
}
```

Where Orleans codegen automatically includes primary constructor arguments before including marked properties. Version tolerance will still be a thing since changing the primary constructor would be a breaking change anyway.

**Reply by ReubenBond** [reply]
This is supported by adding the appropriate attributes to the primary constructor:

```C#
[GenerateSerializer]
record Person([property: Id(0)] string FirstName, [property: Id(1)] string LastName)
{
    [Id(2)]
    public string? MiddleName { get; init; }
}
````

We might be able to imply the properties, but I'm concerned that there would be some corner cases (eg, what if there are multiple constructors?)

**Reply by koenbeuk** [reply]
A record can only have a single primary constructor. The use of additional constructors is possible but (to my experience) not common. If this is desired then you can always explicitly assign Id's to the various properties and fields (as you've demonstrated). 

For our use-case, we use records all the time (intermixed with tuples), often just with a primary constructor, no explicit members, as DTO's. Having to not label each property with an attribute while still keeping the version tolerance benefit would be a great addon. In likely reality the above will likely look like this:
```csharp
record Person(string FirstName, string? MiddleName, string LastName);
```

Any modification to this record either will be a breaking change, or a non-breaking change. The latter should ideally simply involve only applying attributes to those additional properties that we decide to add later on.

**Reply by koenbeuk** [reply]
To emphasize on my last point: If we were to introduce an Age property afterwards through non-breaking manners, the record would ideally look like this:

```csharp
record Person(string FirstName, string? MiddleName, string LastName)
{
    // Optional constructor
    public Person(string firstName, string middleName, string lastName, int age) {          
        FirstName = firstName;
        MiddleName= middleName;
        LastName = lastName;
        Age = age;
    }

    [Id(0)]
    public int? Age { get; init; }
}
```

The minimum amount of work to have both simplicity and version tolerance.

**Reply by ReubenBond** [reply]
I'm ok with this. The [Id(0)] bit is odd, though: I imagine we'd have to error-out if you tried to do that, since the FirstName property would likely have that id. Either that, or the default ids for the constructor properties would have to start at some higher value (eg, 1000). The CodeFix could auto-add the right id for you.

**Reply by koenbeuk** [reply]
I would think that primary constructor properties are automatically assigned an Id (probably in a reserved range as you suggested). Perhaps the negative range can be reserved for primary constructor properties? that way, we can enforce explicit properties to start from 0. Happy to contribute in this area.

**Reply by ReubenBond** [reply]
Negative ids are not valid. Ids starting at 0 and incrementing by 1 are encoded more efficiently, since the difference between subsequent ids are encoded (and properties are encoded in ascending order) and differences of less than 7 between ids can be encoded without consuming extra space (they are encoded directly into the 'tag' byte for the property).

Help is appreciated. If you're willing to tackle this, please feel free to message me on Discord so that I can assist you in case you run into any issue or if you'd like to know where to look

**Reply by VincentH-Net** [reply]
@ReubenBond 
> I'm ok with this. 

I would appreciate this improvement; the current supported functionality with repeated `[property: Id()]` feels like a workaround and unnecessary ceremony; it detracts from the excellent readability of this `record` syntax. 

Having to mark a `record` as `[Immutable]` explicitly seems like unnecessary ceremony as well. I am aware that records are not really guaranteed immutable, but I think it makes more sense for records to be immutable (in the Orleans sense) by default instead of mutable. It will make Orleans applications more performant by default when Co-Hosting - no serialization at all in-process.

I.e. now almost every record in my Orleans project looks like:
```csharp
[GenerateSerializer, Immutable]
public record UserInfo([property: Id(0)] Guid Id, [property: Id(1)] string Name);
```
All those records could be:
```csharp
[GenerateSerializer]
public record UserInfo(Guid Id, string Name);
```
You could theoretically still support mutable records with a `[MutableRecord]` attribute, but even though this is technically possible I think it would be better not to support mutable records. The fact that you would need to apply an attribute means you own the record type, so you also should be able to define the record safe for immutability within Orleans.

Finally - since `[Immutable]` is just an optimization for serialization, and Orleans 4 now has the `[GenerateSerializer]` attribute, would it make more sense to make `[Immutable]` obsolete and instead add an optional `bool? immutable` parameter to `[GenerateSerializer]`? The default would be mutable for classes and immutable for records. So an immutable class would get:
```csharp
[GenerateSerializer(immutable: true)]
public class MyClass
```

**Comment by oising** [comment]
> Having to mark a record as [Immutable] explicitly seems like unnecessary ceremony as well. I am aware that records are not really guaranteed immutable, but I think it makes more sense for records to be immutable (in the Orleans sense) by default instead of mutable. It will make Orleans applications more performant by default when Co-Hosting - no serialization at all in-process.

This is a foot gun. A plain `record` is a reference type, and will lead to disgusting bugs if people forget this. If you want immutable records, use `readonly record struct`. The `[immutable]` attribute is there for a reason - if you're using a byref record and you know what you're doing, then you should have to explicitly opt-in.

**Reply by VincentH-Net** [reply]
@oising using `struct` is only recommended for type instance sizes under 16 bytes ([guideline](https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/choosing-between-class-and-struct)).

In a real-world application only a small part of the `[Immutable]` types would be under 16 bytes, so using `readonly record struct` is not a realistic solution to prevent that type of errors.

I see the point you make; safe `Immutable` by default on `record` would require analyzing the property types for immutability while processing `[GenerateSerializer]` - e.g. `ImmutableArray<T>` for collections.

It could be fairly simple analysis, just have a known list of framework types that can be considered safe for Orleans immutability and only consider records types immutable if they contain no other types. Devs can still override the default either way with an optional `bool? immutable` parameter for the `[GenerateSerializer]` attribute.

I do think that the vast majority of `record` types could be perf optimized automatically for Orleans immutability that way, and also safe. And with less ceremony / noise in the code.
